### PR TITLE
[air] Serialize additional files in dict checkpoints turned dir checkpoints

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

With this PR, files put into directory checkpoints that were dict checkpoints will be serialized and retained when a subsequent `to_dict()` is called. This is to enable storing additional files, as e.g. needed by Ray Tune.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/air/checkpoint.py [hunk 1]
```diff
@@ -48,6 +49,12 @@ class Checkpoint:
     There are no guarantees made about compatibility of intermediate
     representations.
 
+    For checkpoints which were created from dictionaries and then serialized into
+    directories, if additional files are added to the directory, these will be
+    serialized if the checkpoint is subsequently turned into a dictionary again.
+    In this case, the checkpoint dictionary will contain an additional private field,
+    ``_ray_additional_checkpoint_files``, containing the serialized files.
```
Comment (xwjiang2010) [inline python/ray/air/checkpoint.py:56]: Could you describe the behavior in a way that doesn't require users to poke into _ray_additional_checkpoint_files?
Ideally they shouldn't right?
How about let's say dict --> directory (add some more files) --> dict --> directory won't lose the added files?

Also maybe present this paragraph and the next in a coherent way. (they are pretty much the same thing)

File: python/ray/air/checkpoint.py [hunk 2]
```diff
@@ -258,6 +265,23 @@ def to_dict(self) -> dict:
...
+
+                    # If there are additional files in the directory, add them as
+                    # _DICT_CHECKPOINT_ADDITIONAL_FILE_KEY
+                    additional_files = {}
+                    for file_or_dir in os.listdir(local_path):
+                        if file_or_dir in [".", "..", _DICT_CHECKPOINT_FILE_NAME]:
+                            continue
+
+                        additional_files[file_or_dir] = _pack(
+                            os.path.join(local_path, file_or_dir)
```
Comment (Yard1) [inline python/ray/air/checkpoint.py:277]: the path is guaranteed to be relative here right?

File: python/ray/air/checkpoint.py [hunk 3]
```diff
@@ -361,7 +385,15 @@ def _to_directory(self, path: str) -> None:
...
                 _unpack(data_dict[_FS_CHECKPOINT_KEY], path)
             else:
-                # This is a dict checkpoint. Dump data into checkpoint.pkl
+                # This is a dict checkpoint.
+                # First, restore any additional files
+                additional_files = data_dict.pop(
+                    _DICT_CHECKPOINT_ADDITIONAL_FILE_KEY, {}
+                )
+                for file, content in additional_files.items():
+                    _unpack(stream=content, path=os.path.join(path, file))
```
Comment (Yard1) [inline python/ray/air/checkpoint.py:394]: Do we need to create the directories or does tar take care of that?

Comment (krfricke) [reply python/ray/air/checkpoint.py:395]: tar takes care of that

**Review by Yard1 — APPROVED** [review]

**Review by xwjiang2010 — APPROVED** [review]
Thanks Kai!

**Pull request merged**